core-bundle/src/Filesystem/VirtualFilesystem.php

```diff
@@ -0,0 +1,278 @@
+        }
+
+        // Read from DBAFS
+        if (!($accessFlags & self::BYPASS_DBAFS) && $this->dbafsManager->match($path)) {
+            yield from $this->dbafsManager->listContents($path, $deep);
+
+            return;
+        }
+
+        // Read from adapter, but enhance result with extra metadata on demand
+        /** @var FilesystemItem $item */
+        foreach ($this->mountManager->listContents($path, $deep) as $item) {
+            yield $item->withExtraMetadata(
+                fn () => $this->dbafsManager->getExtraMetadata($item->getPath())
+            );
+        }
+    }
+
+    public function getLastModified($location, int $accessFlags = self::NONE): int
+    {
+        $path = $this->resolve($location);
+
+        if ($accessFlags & self::FORCE_SYNC) {
+            $this->dbafsManager->sync($path);
+        }
+
+        if (!($accessFlags & self::BYPASS_DBAFS) && null !== ($lastModified = $this->dbafsManager->getLastModified($path))) {
+            return $lastModified;
+        }
+
+        return $this->mountManager->getLastModified($path);
+    }
+
+    public function getFileSize($location, int $accessFlags = self::NONE): int
+    {
+        $path = $this->resolve($location);
+
+        if ($accessFlags & self::FORCE_SYNC) {
+            $this->dbafsManager->sync($path);
+        }
+
+        if (!($accessFlags & self::BYPASS_DBAFS) && null !== ($fileSize = $this->dbafsManager->getFileSize($path))) {
+            return $fileSize;
+        }
+
+        return $this->mountManager->getFileSize($path);
+    }
+
+    public function getMimeType($location, int $accessFlags = self::NONE): string
+    {
+        $path = $this->resolve($location);
+
+        if ($accessFlags & self::FORCE_SYNC) {
+            $this->dbafsManager->sync($path);
+        }
+
+        if (!($accessFlags & self::BYPASS_DBAFS) && null !== ($mimeType = $this->dbafsManager->getMimeType($path))) {
+            return $mimeType;
+        }
+
+        return $this->mountManager->getMimeType($path);
+    }
+
+    public function getExtraMetadata($location, int $accessFlags = self::NONE): array
+    {
+        $path = $this->resolve($location);
+
+        if ($accessFlags & self::FORCE_SYNC) {
+            $this->dbafsManager->sync($path);
+        }
+
+        if ($accessFlags & self::BYPASS_DBAFS) {
+            return [];
+        }
+
+        return $this->dbafsManager->getExtraMetadata($path);
+    }
+
+    public function setExtraMetadata($location, array $metadata): void
+    {
+        $this->ensureNotReadonly();
+
+        $this->dbafsManager->setExtraMetadata($this->resolve($location), $metadata);
+    }
+
+    /**
+     * @param string|Uuid $location
+     */
+    private function resolve($location): string
+    {
+        if ($location instanceof Uuid) {
+            return $this->dbafsManager->resolveUuid($location, $this->prefix);
+        }
+
+        $path = Path::canonicalize($location);
+
+        if (str_starts_with($path, '..')) {
```

Comment:
Imho that's not secure. I can do `files/../../..` to circumvent this. That's why we have `Validator::isInsecurePath()`. Maybe the Symfony `Path` library can do this for us? /cc @ausi

Reply:
The path is canonicalized at this point, so I think this is fine:

> During normalization, all slashes are replaced by forward slashes ("/").
Furthermore, all "." and ".." segments are removed as far as possible.
".." segments at the beginning of relative paths are not removed.

*(from the Path:canonicalize() doc block)*

Reply:
If we have a testcase for that, I'm fine with it :)

Reply:
Coming up… :smile: